runtime/major_gc.c

```diff
@@ -321,12 +330,11 @@ void caml_final_domain_terminate (caml_domain_state *domain_state)
   }
 }
 
-CAMLno_tsan /* Disable TSan reports from this function (see #11040) */
 static int no_orphaned_work (void)
 {
   return
-    orph_structs.ephe_list_live == 0 &&
-    orph_structs.final_info == NULL;
+    atomic_load_acquire(&orph_structs.ephe_list_live) == 0 &&
```

Comment:
Curious to learn how and why acquire loads are better than relaxed loads here?

Reply:
I'm not a memory-model person so this may be wrong. The intent is that we reliably detect all pushes to orphaning structures that happened before the current point -- instead of observing an arbitrary past value of those lists, which may introduce an arbitrary delay between orphaning and adoption.

Reply:
Does this reasoning work even when the stores are not release stores but instead are protected by an unrelated mutex?

Reply:
On @gadmm's advice I marked the struct field `_Atomic`. I assumed that by default (if we don't specify an ordering) reads and writes are sequentially consistent, so in particular that writes are release stores. (But maybe the default is relaxed? I find it surprisingly difficult to find this information by googling.) The reads in `no_orphaned_work()` are the only ones that are performance-sensitive, so I did not bother specifying an ordering for the others. If we did specify an ordering I would use release stores and acquire reads.

Reply:
>  I marked the struct field _Atomic. I assumed that by default (if we don't specify an ordering) reads and writes are sequentially consistent

I missed this detail, and it makes sense now!

https://en.cppreference.com/w/c/language/atomic

> Built-in [increment and decrement operators](https://en.cppreference.com/w/c/language/operator_incdec) and [compound assignment](https://en.cppreference.com/w/c/language/operator_assignment) are read-modify-write atomic operations with total sequentially consistent ordering (as if using [memory_order_seq_cst](https://en.cppreference.com/w/c/atomic/memory_order)).

Reply:
Thanks. This quote from is about composed operations that both read and write at the same time, not single reads or single writes. But these do also default to sequential consistency, as documented on the *other* page https://en.cppreference.com/w/c/atomic/memory_order :

> The default behavior of all atomic operations in the [language](https://en.cppreference.com/w/c/language/atomic) and the library provides for sequentially consistent ordering (see discussion below). That default can hurt performance, but the library's atomic operations can be given an additional memory_order argument [..].

(I wish this was said clearly on the `atomic` page, but oh well, we have enough work improving the OCaml manual already.)

Reply:
Yes. You are right. I was reading it too quickly.